Mathlib/AlgebraicGeometry/Over.lean

```diff
@@ -0,0 +1,94 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.Scheme
+
+/-!
+# Typeclasses for `S`-schemes and `S`-morphisms
+
+## Main definition
+- `AlgebraicGeometry.Scheme.Over`: `X.Over S` equips `X` with a `S`-scheme structure.
+  `X ⮕ S : X ⟶ S` is the structure morphism.
+- `AlgebraicGeometry.Scheme.Hom.IsOver`: `f.IsOver S` asserts that `f` is a `S`-morphism.
+
+-/
+
+
+namespace AlgebraicGeometry.Scheme
+
+universe u
+
+open CategoryTheory
+
+variable {X Y : Scheme.{u}} (f : X.Hom Y) (S S' : Scheme.{u})
+
+/--
+`X.Over S` is the typeclass containing the data of a structure morphism `X ⮕ S : X ⟶ S`.
+-/
+protected class Over (X S : Scheme.{u}) : Type u where
+  ofHom ::
+  /-- The structure morphism. Use `X ⮕ S` instead. -/
+  hom : X ⟶ S
+
+/--
+The structure morphism `X ⮕ S : X ⟶ S` given `X.Over S`.
+The instance argument is an `optParam` instead so that it appears in the discrimination tree.
+-/
+def over (X S : Scheme.{u}) (_ : X.Over S := by infer_instance) : X ⟶ S := Scheme.Over.hom
+
+/-- The structure morphism `X ⮕ S : X ⟶ S` given `X.Over S`. -/
+notation:90 X:90 " ⮕ " S:90 => Scheme.over X S inferInstance
+
+/-- See Note [custom simps projection] -/
+def Over.Simps.over (X S : Scheme.{u}) [X.Over S] : X ⟶ S := X ⮕ S
+
+initialize_simps_projections Scheme.Over (hom → over)
+
+/--
+`X.CanonicallyOver S` is the typeclass containing the data of a structure morphism `X ⮕ S : X ⟶ S`,
+and that `S` is (uniquely) inferrable from the structure of `X`.
+-/
+class CanonicallyOver (X : Scheme.{u}) (S : semiOutParam (Scheme.{u})) extends X.Over S where
```

Comment:
I am not sure how useful this will be in practice. It's nice that we can have your
```
instance [X.CanonicallyOver Y] [Y.Over S] : X.Over S := _
```
But is that so common? I am guessing that we mostly want to work with `IsOverTower` as an assumption instead as we do in commutative algebra.

Reply:
If `X` is an `S` scheme, we would want various constructions to be `S`-schemes as well e.g. open subschemes, $$\mathrm{Spec}\mathcal{O}_{X, x}$$, affine spaces, base changes to `X` etc. 
You are right that `CanonicallyOver` should not in general appear as assumptions.

Reply:
Hmm, but I think all the examples should work without `CanonicallyOver` (they do work with `Algebra`)

Reply:
Ah nevermind, you mean base change with target `X` not with target `S`. Then indeed there are compositions involved in this case.

Reply:
All examples are all a priori `X`-schemes an not `S`-schemes I believe.

Reply:
I still don't understand how base changes to `X` work. `pullback f g` can equally well be considered as over the left factor and as over the right factor. Do you suggest to choose one as the "canonical" one, as we do for tensor products of algebras?
A similar problem arises with fibers, but here maybe the "canonical" choice of base should be the residue field.

Reply:
1. Yeah. Not ideal but it's the best we could do.
2. We can provide `Over` instances to both, just that only one can be "canonical".